### ISSUE TITLE
Decouple piet and druid-shell

### ISSUE THREAD

**Issue by cmyr** [linebender/druid#1924]
It is not currently possible to use druid-shell without using piet, and I think this is an unnecessary coupling, and it makes `druid-shell` strictly less useful.

I would like to move all of the piet-specific code from `druid-shell` into `druid`. In practice this means:

- Any shared `PietText` objects should live somewhere in druid (unclear where, exactly, druid doesn't really have an application-global object? maybe the `AppHandler`?)
- instead of passing a `RenderContext` to the `paint` method, piet should pass a window handle
- druid should construct the `RenderContext` itself, from the window handle

This does violate the principal that platform-specific code should generally be in druid-shell instead of druid. Is there a viable alternative? I'm not sure if there's a good one, although we could perhaps have a `from_raw_handle(handle: &RawWindowHandle) -> RenderContext` method in `piet-common`? I'm not sure if this really makes a ton of sense, but it's an option?

If someone is interested in taking this on, I am happy to answer questions.

**Comment by JAicewizard** [comment]
> instead of passing a RenderContext to the paint method, piet should pass a window handle
piet-> druid-shell?

I am not sure how easy it is to decouple piet and druid-shell. GTK3 only has cairo/some ogl support, gtk4 has better GPU support, but in the end we're limited by these fixed backends.

Maybe its an idea to have some kind of thing that allows druid-shell to provide an interface that can be rendered to. So that a widget can draw into the a surface, and druid-shell then draws the final result to the cairo-surface.

Like an efficient version of `make_image`+`draw_image` for, for example, ogl. Like `make_ogl`+`draw_ogl` that can be efficienlty projected onto the `RenderContext` without converting image data too much.

**Comment by cmyr** [comment]
> > instead of passing a RenderContext to the paint method, piet should pass a window handle
> 
> piet-> druid-shell?

Oops, yes. fixed. :)

> I am not sure how easy it is to decouple piet and druid-shell. GTK3 only has cairo/some ogl support, gtk4 has better GPU support, but in the end we're limited by these fixed backends.
> 
> Maybe its an idea to have some kind of thing that allows druid-shell to provide an interface that can be rendered to. So that a widget can draw into the a surface, and druid-shell then draws the final result to the cairo-surface.
> 
> Like an efficient version of `make_image`+`draw_image` for, for example, ogl. Like `make_ogl`+`draw_ogl` that can be efficienlty projected onto the `RenderContext` without converting image data too much.

I don't want to limit druid-shell based on the capacities of the most limited backend, and I also don't want to over-complicate the API; I think consumers can decide what makes the most sense for them (such as only supporting x11?).

Another way to think about this: everything that is currently possible remains possible, and some set of things that were not possible become possible?

**Comment by JAicewizard** [comment]
> I don't want to limit druid-shell based on the capacities of the most limited backend, and I also don't want to over-complicate the API; I think consumers can decide what makes the most sense for them (such as only supporting x11?).

Yes I know, but I assume there are similar limitations on other platforms. I also dont think this would be a limmiting API, are there any usecases that wouldnt be covered by this? besides using an unsupported API.

**Comment by cmyr** [comment]
As far as I understand it, these limitations are not present on other platforms. I would basically point to the winit crate: we should generally try to do what they do (although they do not have a GTK backend).

I'm also not sure what you mean by limiting the API? If someone wants to make a game using openGL and they want to build that on top of druid-shell this is not possible if piet is bundled.

**Comment by Maan2003** [comment]
> This does violate the principal that platform-specific code should generally be in druid-shell instead of druid. Is there a viable alternative?

We can have `WindowHandle::get_render_context` but behind a feature `"piet"`

Also I think it is _OK_ if gtk backend does not compile if `"piet"` feature is not enabled and druid-shell defaults to x11 backend

**Comment by cmyr** [comment]
yes, that would probably work. One issue is where the text state should be stored, but I'm sure that's solvable.

**Comment by JAicewizard** [comment]
> I'm also not sure what you mean by limiting the API? If someone wants to make a game using openGL and they want to build that on top of druid-shell this is not possible if piet is bundled.

You said that you dont want to limit the API on one backend, but I dont think what I outlined limits what people can do. We can let the user draw using OGL, and then we can paint their rendered image onto the cairo surface. I dont know exactly what this API should look like but it should be achievable.

**Comment by cmyr** [comment]
> You said that you dont want to limit the API on one backend, but I dont think what I outlined limits what people can do. We can let the user draw using OGL, and then we can paint their rendered image onto the cairo surface. I dont know exactly what this API should look like but it should be achievable.

This would require extracting the frame from the GPU as a bitmap and then drawing that bitmap into a cairo context. This might be possible, but it would be hugely inefficient and a ton of additional work for minimal obvious gain. I think we should keep things simple, and not worry *overly* much about compatibility: my goal is just to remove limitations, and then users of druid-shell can experiment however they want. :)
